File: repl/adrianus/cpp/lexer.cpp
#include "lexer.h"
#include "token.cpp"
#include <iostream>

Lexer::Lexer() {

}

Lexer::~Lexer() {

}

void Lexer::Load(std::string s) {
    source = s;
    position = 0;
    readPosition = 0;
    ReadChar();
}

void Lexer::ReadChar() {
    if (readPosition < source.length()) {
        current_char = source[readPosition];
    } else {
        current_char = 0;
    }
    position = readPosition;
    ++readPosition;
}

void Lexer::SkipWhitespace() {
    while (current_char == ' ' || current_char == '\t' || current_char == '\r' || current_char == '\n') {
        ReadChar();
    }
}

bool Lexer::IsLetter() {
    return (('a' <= current_char) && (current_char <= 'z')) || (('A' <= current_char) && (current_char <= 'Z')) || current_char == '_';
}

bool Lexer::IsDigit() {
    return ('0' <= current_char) && (current_char <= '9');
}

std::string Lexer::ReadIdentifier() {
    int start = position;
    while(IsLetter() || IsDigit()) {
        ReadChar();
    }
    return source.substr(start, position - start);
}

std::string Lexer::ReadNumber() {
    int start = position;
    while(IsDigit()) {
        ReadChar();
    }
    if (IsFloatDot()) {
        ReadChar();
        while(IsDigit()) {
            ReadChar();
        }
    }
    return source.substr(start, position - start);
}

TokenType Lexer::LookupIdent(std::string ident) {
    if (keywords.find(ident) != keywords.end()) {
        return keywords.at(ident);
    }
    return TT_IDENT;
}

std::string Lexer::ReadDoubleQuotesString() {
    ReadChar();
    int start = position;
    while (current_char != '"') {
        if (current_char == '\\' && PeekChar() == '"') {  // escaping \"
            ReadChar();
        }
        ReadChar();
    }
    return source.substr(start, position - start);
}

std::string Lexer::ReadSingleQuotesString() {
    ReadChar();
    int start = position;
    while (current_char != '\'') {
        if (current_char == '\\' && PeekChar() == '\'') {  // escaping \'
            ReadChar();
        }
        ReadChar();
    }
    return source.substr(start, position - start);
}

char Lexer::PeekChar() {
    return source[readPosition];
}

bool Lexer::IsEOF() {
    return current_char == 0;
}

bool Lexer::IsFloatDot() {
    return current_char == '.';
}

Token Lexer::NextToken() {
    Token token;
    SkipWhitespace();
    bool do_read_char_after_lexing = true;
    switch(current_char) {
        case 0:
            token.type = TT_EOF;
            token.literal = "";
        break;
        case '+':
            if (PeekChar() == '+') {
                ReadChar();
                token.type = TT_PLUSPLUS;
                token.literal = "++";
            } else {
                token.type = TT_PLUS;
                token.literal = current_char;
            }
        break;
        case '-':
            token.type = TT_MINUS;
            token.literal = current_char;
        break;
        case '*':
            token.type = TT_ASTERISK;
            token.literal = current_char;
        break;
        case '%':
            token.type = TT_MODULUS;
            token.literal = current_char;
        break;
        case '/':
            if (PeekChar() == '*') {
                ReadChar();
                token.type = TT_MULTICOMMENT;
                token.literal = "/*";
                while (!(current_char == '*' && PeekChar() == '/')) {
                    ReadChar();
                }
                ReadChar();
                ReadChar();
            } else if (PeekChar() == '/') {
                token.type = TT_SINGLECOMMENT;
                token.literal = "//";
                ReadChar();
                while (current_char != '\n' && current_char != 0) {
                    ReadChar();
                }
            } else {
                token.type = TT_SLASH;
                token.literal = current_char;
            }
        break;
        case '<':
            if (PeekChar() == '=') {
                ReadChar();
                token.type = TT_LTE;
                token.literal = "<=";
            } else {
                token.type = TT_LT;
                token.literal = current_char;
            }
        break;
        case '>':
            if (PeekChar() == '=') {
                ReadChar();
                token.type = TT_GTE;
                token.literal = ">=";
            } else {
                token.type = TT_GT;
                token.literal = current_char;
            }
        break;
        case ',':
            token.type = TT_COMMA;
            token.literal = current_char;
        break;
        case ';':
            token.type = TT_SEMICOLON;
            token.literal = current_char;
        break;
        case '(':
            token.type = TT_LPAREN;
            token.literal = current_char;
        break;
        case ')':
            token.type = TT_RPAREN;
            token.literal = current_char;
        break;
        case '{':
            token.type = TT_LBRACE;
            token.literal = current_char;
        break;
        case '}':
            token.type = TT_RBRACE;
            token.literal = current_char;
        break;
        case ':':
            token.type = TT_COLON;
            token.literal = current_char;
        break;
        case '"':
            token.type = TT_DOUBLE_QUOTES;
            token.literal = ReadDoubleQuotesString();
        break;
        case '\'':
            token.type = TT_SINGLE_QUOTES;
            token.literal = ReadSingleQuotesString();
        break;
        case '[':
            token.type = TT_LBRACKET;
            token.literal = current_char;
        break;
        case ']':
            token.type = TT_RBRACKET;
            token.literal = current_char;
        break;
        case '.':
            token.type = TT_DOT;
            token.literal = current_char;
        break;
        case '=':
            if (PeekChar() == '=') {
                ReadChar();
                token.type = TT_EQ;
                token.literal = "==";
            } else {
                token.type = TT_ASSIGN;
                token.literal = current_char;
            }
        break;
        case '!':
            if (PeekChar() == '=') {
                ReadChar();
                token.type = TT_NOT_EQ;
                token.literal = "!=";
            } else {
                token.type = TT_BANG;
                token.literal = current_char;
            }
        break;
        default:
            if (IsLetter()) {
                std::string ident = ReadIdentifier();
                token.type = LookupIdent(ident);
                token.literal = ident;
                do_read_char_after_lexing = false;
            } else if (IsDigit()) {
                std::string num = ReadNumber();
                if (num.find(".") != std::string::npos) {
                    token.type = TT_FLOAT;
                } else {
                    token.type = TT_INT;
                }
                token.literal = num;
                do_read_char_after_lexing = false;
            } else {
                token.type = TT_ILLEGAL;
                token.literal = "";
            }
        break;
    }
    if (do_read_char_after_lexing) ReadChar();
    return token;
}

File: repl/adrianus/python/lexer.py
from token import Token, TokenType
from token import keywords_map, token_type_map


class Lexer(object):
    def __init__(self, position=0, read_position=0, source=''):
        self.position = position
        self.read_position = read_position
        self.ch = ''
        self.source = source
        self.read_char()

    def load(self, source):
        self.position = 0
        self.read_position = 0
        self.ch = ''
        self.source = source
        self.read_char()

    def next_token(self):
        token = Token()
        self.skip_whitespaces()
        skip_read_char = False
        if self.ch == '+':
            token.type = TokenType.PLUS
            token.literal = self.ch
        elif self.ch == '-':
            token.type = TokenType.MINUS
            token.literal = self.ch
        elif self.ch == '/':
            token.type = TokenType.SLASH
            token.literal = self.ch
        elif self.ch == '*':
            token.type = TokenType.ASTERISK
            token.literal = self.ch
        elif self.ch == '>':
            if self.peek_char() == '=':
                self.read_char()
                token.type = TokenType.GTE
                token.literal = ">="
            else:
                token.type = TokenType.GT
                token.literal = self.ch
        elif self.ch == '<':
            if self.peek_char() == '=':
                token.type = TokenType.GTE
                token.literal = "<="
            else:
                token.type = TokenType.LT
                token.literal = self.ch
        elif self.ch == ';':
            token.type = TokenType.SEMICOLON
            token.literal = self.ch
        elif self.ch == '(':
            token.type = TokenType.LPAREN
            token.literal = self.ch
        elif self.ch == ')':
            token.type = TokenType.RPAREN
            token.literal = self.ch
        elif self.ch == '{':
            token.type = TokenType.LBRACE
            token.literal = self.ch
        elif self.ch == '}':
            token.type = TokenType.RBRACE
            token.literal = self.ch
        elif self.ch == ',':
            token.type = TokenType.COMMA
            token.literal = self.ch
        elif self.ch == 0:
            token.type = TokenType.EOF
            token.literal = self.ch
        elif self.ch == '"':
            token.type = TokenType.STRING
            token.literal = self.read_string()
        elif self.ch == '=':
            if self.peek_char() == '=':
                self.read_char()
                token.type = TokenType.EQ
                token.literal = '=='
            else:
                token.type = TokenType.ASSIGN
                token.literal = self.ch
        elif self.ch == '!':
            if self.peek_char() == '=':
                self.read_char()
                token.type = TokenType.NOT_EQ
                token.literal = '!='
            else:
                token.type = TokenType.BANG
                token.literal = '!'
        else:
            if self.is_letter():
                token.literal = self.read_ident()
                token.type = self.lookup_ident(token.literal)
                skip_read_char = True
            elif self.is_digit():
                token.literal = self.read_number()
                token.type = TokenType.INT
                skip_read_char = True
            else:
                token.type = TokenType.ILLEGAL
                token.literal = ''
        if not skip_read_char:
            self.read_char()
        return token

    def read_char(self):
        if self.read_position < len(self.source):
            self.ch = self.source[self.read_position]
        else:
            self.ch = 0
        self.position = self.read_position
        self.read_position += 1

    def skip_whitespaces(self):
        while self.ch == ' ' or self.ch == '\t' or self.ch == '\r' or self.ch == '\n':
            self.read_char()

    def is_letter(self):
        return ('a' <= self.ch and self.ch <= 'z') or ('A' <= self.ch and self.ch <= 'Z') or self.ch == '_'

    def is_digit(self):
        return '0' <= self.ch and self.ch <= '9'

    def lookup_ident(self, ident):
        if ident in keywords_map:
            return keywords_map[ident]
        return TokenType.IDENT

    def peek_char(self):
        if self.read_position >= len(self.source):
            return 0
        return self.source[self.read_position]

    def read_ident(self):
        start = self.position
        while self.is_letter():
            self.read_char()
        return self.source[start:self.position]

    def read_number(self):
        start = self.position
        while self.is_digit():
            self.read_char()
        return self.source[start:self.position]

    def read_string(self):
        self.read_char()
        start = self.position
        while self.ch != '"':
            self.read_char()
        return self.source[start:self.position]